### PR TITLE
Add support for the WT_CURSOR.equals method.

### PR DESCRIPTION
**Pull request by keithbostic**
@michaelcahill, @agorrod: here's support for WT_CURSOR.compare_equal.

I made the obvious change for Java documentation, but I have not built the Java API (my Java build is busted for some reason I can't figure out).

I think this is ready to merge once it's reviewed and the Java changes added.

### CODE REVIEW

File: src/btree/bt_cursor.c [hunk 1]
```diff
@@ -867,6 +870,42 @@ __cursor_equals(WT_CURSOR_BTREE *a, WT_CURSOR_BTREE *b)
...
+	 * method.
+	 */
+	if (F_ISSET(a, WT_CURSTD_KEY_INT) && F_ISSET(b, WT_CURSTD_KEY_INT))
+		*cmpp = !__cursor_equals(a_arg, b_arg);
+	else {
+		WT_RET(__wt_btcur_compare(a_arg, b_arg, cmpp));
+		if (*cmpp != 0)
+			*cmpp = 1;
+	}
+	return (0);
```
Comment (keithbostic) [inline src/btree/bt_cursor.c]: @michaelcahill, this is one thing I'm a little concerned about -- we have to know if the cursors reference tree positions before taking the fast path, and I'm using `WT_CURSTD_KEY_INT` to figure that out.

**Comment by wtbuild** [comment]
Test PASSed.

**Comment by agorrod** [comment]
@keithbostic Did you consider dropping the cmpp parameter and using WT_NOTFOUND for a mismatch? That would save applications creating a local variable, and means the function can be called inline in conditional statements.

**Comment by keithbostic** [comment]
@agorrod: I made the signatures of `WT_CURSOR.compare` and `WT_CURSOR.compare_equals` identical for two reasons: first, consistency, I like it that both compare functions look the same.

Second, if there's no separate `compare_equals` cursor code path, or if there is a separate `compare_equals` code path but this particular cursor pair can't use it (for example, the btree cursor key is set but not positioned), it's trivial to fall back to `WT_CURSOR.compare` without additional work.

I don't feel terribly strongly about those reasons, we could code around the second reason inside WiredTiger if we wanted.

File: src/include/wiredtiger.in [hunk 2]
```diff
@@ -270,7 +270,8 @@ struct __wt_cursor {
 	 */
 	/*!
 	 * Return the ordering relationship between two cursors: both cursors
-	 * must have the same data source and have valid keys.
+	 * must have the same data source and have valid keys. (When testing
+	 * only for equality, WT_CURSOR::compare_equality may be faster.)
```
Comment (sueloverso) [inline src/include/wiredtiger.in]: That should be WT_CURSOR::compare_equal.  Is there other documentation describing return values or if/when there is any reason a user may want to use compare instead for anything other than less than or greater than now?  Does this just return 0 or non-zero?

Comment (keithbostic) [inline src/include/wiredtiger.in]: Thanks, fixed.

There's no other documentation here in WT_CURSOR::compare (there's separate WT_CURSOR::compare_equal documentation, of course).

WT_CURSOR::compare_equal returns 0/!0 in the parameter, the function returns error status.

**Comment by wtbuild** [comment]
Test PASSed.

**Comment by michaelcahill** [comment]
Does anyone object to renaming this `WT_CURSOR::equals`, which is the normal name for iterators?

File: src/btree/bt_cursor.c (hunk 1 above)
Comment (michaelcahill) [inline src/btree/bt_cursor.c]: I'm missing something: why do we ever need to take the slow path?  That is, if either cursor isn't positioned, they can't be equal by definition, right?

Or are you trying to cover a case where keys are set equal but the cursors are never used for anything?  I'd be more inclined to make it an error if either cursor isn't positioned...

**Comment by sueloverso** [comment]
@michaelcahill does the name change imply changing the signature per @agorrod's suggestion as well so that it can be used directly in a conditional?  Or if not, will the naming confuse people into thinking it can be used that way?

**Comment by keithbostic** [comment]
No objection from me to the rename -- I prefer the current signature as noted above.

File: src/btree/bt_cursor.c (hunk 1 above)
Comment (keithbostic) [inline src/btree/bt_cursor.c]: If the keys are set (but not positioned), they're still equal.

I don't see a reason to force failure if a cursor isn't positioned, it's an odd semantic and differs from WT_CURSOR.compare unnecessarily?

**Comment by wtbuild** [comment]
Test FAILed.

**Comment by wtbuild** [comment]
Test PASSed.

**Comment by wtbuild** [comment]
Test PASSed.

File: src/btree/bt_cursor.c [hunk 3]
```diff
@@ -867,6 +870,43 @@ __cursor_equals(WT_CURSOR_BTREE *a, WT_CURSOR_BTREE *b)
...
+	b = (WT_CURSOR *)b_arg;
+	session = (WT_SESSION_IMPL *)a->session;
+
+	/* Confirm both cursors reference the same object. */
+	if (a_arg->btree != b_arg->btree)
+		WT_RET_MSG(
+		    session, EINVAL, "Cursors must reference the same object");
+
+	/*
+	 * The reason for a compare-equal method is because we can avoid doing
```
Comment (sueloverso) [inline src/btree/bt_cursor.c]: compare-equal should be equals in the comment.

**Comment by wtbuild** [comment]
Test PASSed.

**Comment by michaelcahill** [comment]
Hi @keithbostic, lets talk through 76addf7 -- having "equality == 0" was making my head hurt.

**Comment by wtbuild** [comment]
Test PASSed.

**Comment by sueloverso** [comment]
> having "equality == 0" was making my head hurt.

I suspect that this comes from consistency with strcmp, LSN compare and cursor->compare.

**Comment by michaelcahill** [comment]
Sure, but python tests were asserting `c1->equals(c2) == 0` when the cursors are equal.  We've had an internal `__cursor_equals` for a long time that returns 1 for equality: we were flipping that return to implement the public version.

**Comment by wtbuild** [comment]
Test PASSed.

**Comment by wtbuild** [comment]
Test PASSed.

**Pull request merged**